### DISCUSSION TITLE
[Q] Header Validation with Pydantic Annotaded and AfterValidator

### DISCUSSION THREAD

**Discussion by FSchmidt-FUNKE** [fastapi/fastapi#11068] [Questions]
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/pydantic/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
from pydantic.functional_validators import Annotated, AfterValidator
from fastapi import Header, FastAPI, HTTPException

app = FastAPI()

def is_valid(cost_tracking_id):
    if cost_tracking_id.split('#')[0] in ["intern", "extern"]:
        return cost_tracking_id

    print("Please provide a valid X-Cost-Tracking-ID header.")

def cost_tracking_header_factory():
    return "N/A"

cost_tracking_header = Header(
    description='Includes the cost tracking id',
    default_factory=cost_tracking_header_factory
)

CostTrackingType = Annotated[Annotated[str, AfterValidator(is_valid)], cost_tracking_header]

@app.get("/")
def root(X_Cost_Tracking_ID: CostTrackingType):
    """
    Handler function for the root endpoint.

    Returns:
        str: A string with the value of the X-Cost-Tracking-ID header.
    """
    return X_Cost_Tracking_ID
```

### Description

If I put a breakpoint on the return of the cost_tracking_header_factory and a breakpoint into the first line of the is_valid function. After that I start the server in debugging mode and call the endpoint, without setting any specific headers. When doing that the first stop is at the cost_tracking_header_factory function. Which seems logical to me, because to validate the header I need to read it's value first. Due to the fact that I have not provided a header, the default factory responds and returns "N/A". I would now expect to stop at the breakpoint in the is_valid function next. However, this never happens. Can someone explain to me exactly what happens in the background that the setting of the header value via the default factory is ignored for further validation steps?

### Operating System

Linux, Windows

### Operating System Details

I use PyCharm on Windows and my Python interpreter runs on Ubuntu WSL

### FastAPI Version

0.100.0

### Pydantic Version

2.5.3

### Python Version

3.11.4

**Comment by YuriiMotov** [comment]
This was fixed in FastAPI 0.112.4 and now works as expected! 🎉
